lib/providers/vault.ex

```diff
@@ -0,0 +1,115 @@
+defmodule Ptolemy.Providers.Vault do
+  @moduledoc """
+  A module that provides Vault secrets.
+  
+  # Example
+  Add to your configuration:
+  ```elixir
+  alias Ptolemy.Providers.SystemEnv
+  config :ptolemy, loader: [
+    env: [
+      {{:app_name, :config_key}, {Vault, "VAR_NAME"}}
+    ]
+  ]
+  ```
+  The `:app_name` and `:config_key` are added to reference the application environment
+  key that will be set. The value can be accessed at any time with `Application.get_env/2`.
+  `"VAR_NAME"` is the name of the system environment variable that will be loaded on
+  application startup. Loaded values are all static and will never be updated during runtime.
+
+  # Required Environment Variables
+  None
+  """
+  
+  use Ptolemy.Provider
+
+  @doc """
+  init/1 starts Ptolemy servers are supervised processes under current running process. It 
+  tries to start all the servers configured in config.exs. If all the servers are started
+  successfully, it return :ok back. Else, it returns {:error, res} with res being a list of
+  failed servers.
+  """
+  @callback init(pid) :: :ok | {:error, String.t()}
+  def init(_loader_pid) do
+    res = Application.get_env(:ptolemy, :vaults)
+      |> Enum.map(fn {k, _} -> k end)
+      |> Enum.reject(&(elem(Ptolemy.start(&1, &1), 0) == :ok))
+
+    case Enum.empty?(res) do
+        true -> :ok
+        false -> {:error, res}
+    end
+  end
+
+  @doc """
+  load/2 reads in the data from Ptolemy's vault interface and check if the secret will expire.
+  If it does, it would register the secret's ttl with the loader, so the loader can load the
+  secret again later. If the ttl doesn't exist, it will skip and return the data.
+
+  var_args is a 4-element list
```

Comment:
Why not typespec this arg?

Reply:
Good idea! I will add it.

Reply:
Elixir typespec does not support specifying something like [pid(), atom(), [any()], [any()]], so i only included types that should be allowed in the list.